lib/AST/Module.cpp

```diff
@@ -2438,28 +2438,36 @@ bool SourceFile::hasTestableOrPrivateImport(
       });
 }
 
-bool SourceFile::isImportedImplementationOnly(const ModuleDecl *module) const {
-  // Implementation-only imports are (currently) always source-file-specific,
-  // so if we don't have any, we know the search is complete.
-  if (!hasImplementationOnlyImports())
-    return false;
-
+RestrictedImportKind SourceFile::getRestrictedImportKind(const ModuleDecl *module) const {
   auto &imports = getASTContext().getImportCache();
+  RestrictedImportKind importKind = RestrictedImportKind::Implicit;
 
   // Look at the imports of this source file.
   for (auto &desc : *Imports) {
     // Ignore implementation-only imports.
-    if (desc.options.contains(ImportFlags::ImplementationOnly))
+    if (desc.options.contains(ImportFlags::ImplementationOnly)) {
+      if (imports.isImportedBy(module, desc.module.importedModule))
+        importKind = RestrictedImportKind::ImplementationOnly;
       continue;
+    }
 
-    // If the module is imported this way, it's not imported
+    // If the module is imported publicly, it's not imported
     // implementation-only.
     if (imports.isImportedBy(module, desc.module.importedModule))
-      return false;
+      return RestrictedImportKind::None;
   }
 
   // Now check this file's enclosing module in case there are re-exports.
-  return !imports.isImportedBy(module, getParentModule());
+  if (imports.isImportedBy(module, getParentModule()))
+    return RestrictedImportKind::None;
+
+  // Special case the __ObjC module. This module is implicitly imported but
+  // represents any clang module so we can consider it public.
```

Comment:
I'm having trouble following this part because I thought the `__ObjC` module is the one that gets implicitly imported as a result of having a bridging header. If that's true, I would think that it is harmful to reference decls from `__ObjC` in public decls and inlinable code, at least for library evolution enabled modules. Am I thinking about this wrong?

Reply:
I though frameworks aren't meant to have bridging headers at all, especially not if they're built with library evolution.

Reply:
I experimented with it and it looks like there's an automatic `@_exported import __ObjC` emitted in the swiftinterface file for a module with a bridging header, so `__ObjC` is implicitly publicly imported. Using these decls in API should be fine in theory. This appeared to work even without that special case, I believe the import of a bridging header `__ObjC` is usually recognized by the check a few lines above: `if (imports.isImportedBy(module, desc.module.importedModule))`

I actually added this special case to support the how the C++ interop represents C++ namespaces in Swift. They synthesize an enum in the `__ObjC` module and extend it in each imported module adding decls to the namespace. This lead to this diagnostics reporting references to `__ObjC` as errors in the Swift code in the compiler for each use of a decl under a namespace. There's a few more issues here, the first one being that extension members can be referenced without being imported so the C++ interop solution suffers greatly from the very bug this PR tries to limit. As a proper alternative to this special case, maybe the C++ interop should instead declare the import more cleanly as we already do for a module with a bridging header.

Reply:
Eventually I'd like to fix that, though. I think the reason it's discouraged is what I'm talking about here, but if we got the diagnostics correct and changed the semantics for library evolution modules to have the bridging header be implicitly implementation only imported I think we we could allow it. Many framework developers need something like a bridging header but have to resort to creating a custom clang module to import implementation only instead.

Reply:
(Not sure how the comment order got mixed up, but my last comment was in response to Slava's).

How would the compiler resolve the `import __ObjC` when encountering it in a swiftinterface in an SDK? It seems like that would rely on the corresponding clang header being available at some well known location relative to the swiftmodule. In an Xcode project a bridging header is usually just a project header, though. And if you didn't make it a project header, I think the build system would need to know to treat it specially and synthesize a module definition file to install with it or something.

Reply:
I don't know how building that swiftinterface works in practice, like Slava I'm used to Xcode (I believe it's Xcode) rejecting frameworks with bridging-headers. Thinking about it, the use case with the C++ interop in the compiler was for a mixed source module, I'll have to take a look at how it behaves in such a case.

Reply:
I'll remov the special case for now to keep the check as strict as it was before, even if it may raise more warnings now as it's executed on more files.